eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs

```diff
@@ -396,16 +396,26 @@ ledgerTransition = do
               , certState
               , StrictSeq.fromStrict $ txBody ^. certsTxBodyL
               )
-        let wdrlAddrs = Map.keysSet . unWithdrawals $ tx ^. bodyTxL . withdrawalsTxBodyL
-            wdrlCreds = Set.map raCredential wdrlAddrs
-            dUnified = dsUnified $ certDState certStateAfterCERTS
-            delegatedAddrs = DRepUView dUnified
-
-        -- TODO: Finish this implementation once we are in bootstrap phase:
-        -- https://github.com/IntersectMBO/cardano-ledger/issues/4092
-        when False $ do
-          all (`UMap.member` delegatedAddrs) wdrlCreds
-            ?! ConwayWdrlNotDelegatedToDRep (wdrlCreds Set.\\ Map.keysSet (dRepMap dUnified))
+
+        let checkWdrlsKeyHashesDelegated = do
+              let
+                isKeyHash = \case
+                  KeyHashObj _ -> True
+                  _ -> False
+                wdrlsKeyHashes =
+                  Set.filter isKeyHash $
+                    Set.map raCredential $
+                      Map.keysSet $
+                        unWithdrawals $
+                          tx ^. bodyTxL . withdrawalsTxBodyL
+                dUnified = dsUnified $ certDState certStateAfterCERTS
+                notDelegated = wdrlsKeyHashes Set.\\ Map.keysSet (dRepMap dUnified)
+              Set.null notDelegated ?! ConwayWdrlNotDelegatedToDRep notDelegated
+
+        -- Starting with version 10, we don't allow withdrawals into RewardAcounts that are KeyHashes and not delegated to Dreps
+        unless
+          (HF.bootstrapPhase (pp ^. ppProtocolVersionL))
```

Comment:
Don't do a separate binding. Those binding for non-reusable functions only introduce unnecessary hops when reading the code
```suggestion
        -- Starting with version 10, we don't allow withdrawals into RewardAcounts that are KeyHashes and not delegated to Dreps
        unless (HF.bootstrapPhase (pp ^. ppProtocolVersionL)) $ do
          ... inline here the whole binding
```